### ISSUE TITLE
Convenience methods for range-check operations

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/portable-simd#275]
So, I find myself often doing something like `val.lanes_ge(start) & val.lanes_le(end)` and was wondering how well an alternative like `val.lanes_range(start..=end)` would fit on the provided SIMD types. I don't know of any platform which would provide this as a built-in operation, and it would be equivalent to just running the operations in sequence.

I would offer a PR for this, but I mostly ask since I'm not sure if it would be reasonable to offer "convenience" operations like this instead of purely ones based upon actual SIMD instructions. To fully clarify, the implementation would be effectively:

```rust
use core::ops::{RangeBounds, Bound};

#[inline]
fn simd_range(self, other: impl RangeBounds<Self>) -> Self::Mask
{
    let start = match other.start_bound() {
        Bound::Included(&start) => Some(self.simd_ge(start)),
        Bound::Excluded(&start) => Some(self.simd_gt(start)),
        Bound::Unbounded => None,
    };
    let end = match other.end_bound() {
        Bound::Included(&end) => Some(self.simd_le(end)),
        Bound::Excluded(&end) => Some(self.simd_lt(end)),
        Bound::Unbounded => None,
    };
    match (start, end) {
        (Some(start), Some(end)) => start & end,
        (None, Some(one)) | (Some(one), None) => one,
        (None, None) => Self::Mask::splat(true),
    }
}
```

Which would (presumably) optimise down to just `self.simd_ge(start) & self.simd_le(end)` for `self.simd_range(start..=end)`, but I haven't tested to verify this claim. The `self.simd_range(..)` seems strange to support but `RangeBounds` doesn't explicitly deny it.

**Comment by workingjubilee** [comment]
While it's not a priority by any means, I don't think a convenience like this is out of the question. There are some operations which we offer that are almost purely programmatic convention and have no hardware equivalent, and others which we offer not because the hardware can do better, but because the compiler can.

**Comment by calebzulawski** [comment]
What about providing a `SimdRangeBound` trait, implemented over all of the ranges in `std::ops`?

**Comment by clarfonthey** [comment]
I'm not sure why you'd need to create a separate trait for this.

**Comment by calebzulawski** [comment]
The intention there is to keep it similar to the scalar API.  It's not clear to me that this function should be implemented on vectors themselves.

**Comment by clarfonthey** [comment]
I see, you're saying that they should be inherent methods on the ranges, but since this crate can't do that, until it's merged into libstd, we should provide an extension trait instead.

**Comment by calebzulawski** [comment]
Yeah, more or less.  Though `RangeBounds` is already a trait, I was suggesting implementing a trait over that, rather than on the ranges themselves.
